pysd/py_backend/output.py

```diff
@@ -0,0 +1,513 @@
+        """
+        Writes values of variables in capture_elements in the netCDF4 Dataset.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+        """
+        for key in capture_elements:
+
+            comp = model[key]
+
+            if "time" in self.ds[key].dimensions:
+                if isinstance(comp, xr.DataArray):
+                    self.ds[key][self.step, :] = comp.values
+                else:
+                    self.ds[key][self.step] = comp
+            else:
+                try:  # this issue can arise with external objects
+                    if isinstance(comp, xr.DataArray):
+                        self.ds[key][:] = comp.values
+                    else:
+                        self.ds[key][:] = comp
+                except ValueError:
+                    warnings.warn(
+                        f"The dimensions of {key} in the results "
+                        "do not match the declared dimensions for this "
+                        "variable. The resulting values will not be "
+                        "included in the results file.")
+
+        self.__update_step()
+
+    def postprocess(self, **kwargs):
+        """
+        Closes netCDF4 Dataset.
+
+        Returns
+        -------
+        None
+        """
+
+        # close Dataset
+        self.ds.close()
+
+        print(f"Results stored in {self.out_file}")
+
+    def add_run_elements(self, model, capture_elements):
+        """
+        Adds constant elements to netCDF4 Dataset.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: list
+            List of constant elements
+
+        Returns
+        -------
+        None
+        """
+        # creating variables in capture_elements
+        self.__create_ds_vars(model, capture_elements, time_dim=False)
+
+        self.update(model, capture_elements)
+
+    def __create_ds_vars(self, model, capture_elements, time_dim=True):
+        """
+        Create new variables in a netCDF4 Dataset from the capture_elements.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+        time_dim: bool
+            Whether to add time as the first dimension for the variable.
+
+        Returns
+        -------
+        None
+
+        """
+        for key in capture_elements:
+            comp = model[key]
+
+            dims = ()
+
+            if isinstance(comp, xr.DataArray):
+                dims = tuple(comp.dims)
+
+            if time_dim:
+                dims = ("time",) + dims
+
+            var = self.ds.createVariable(key, "f8", dims, compression="zlib")
```

Comment:
The compression is not available in netcdf4 1.5.0. Could you make an if/else block for being able to keep the support for Python 3.7 in Windows

Reply:
Done, on Monday I will try to think on ways to test this